plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -0,0 +1,568 @@
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+module Language.UntypedPlutusCore.Evaluation.Machine.Cek
+    ( CekValue(..)
+    , CekEvaluationException
+    , EvaluationResult(..)
+    , ErrorWithCause(..)
+    , EvaluationError(..)
+    , ExBudget(..)
+    , ExBudgetCategory(..)
+    , ExBudgetMode(..)
+    , ExRestrictingBudget(..)
+    , ExTally(..)
+    , CekExBudgetState
+    , CekExTally
+    , exBudgetStateTally
+    , extractEvaluationResult
+    , runCek
+    , runCekCounting
+    , evaluateCek
+    , unsafeEvaluateCek
+    , readKnownCek
+    )
+where
+
+import           PlutusPrelude
+
+import           Language.UntypedPlutusCore.Core
+import           Language.UntypedPlutusCore.Subst
+
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Error
+import           Language.PlutusCore.Evaluation.Machine.ExBudgeting
+import           Language.PlutusCore.Evaluation.Machine.Exception
+import           Language.PlutusCore.Evaluation.Machine.ExMemory
+import           Language.PlutusCore.Evaluation.Result
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Pretty
+import           Language.PlutusCore.Universe
+
+import           Control.Lens                                       (AReview)
+import           Control.Lens.Operators
+import           Control.Lens.Setter
+import           Control.Monad.Except
+import           Control.Monad.Morph
+import           Control.Monad.Reader
+import           Control.Monad.State.Strict
+import           Data.Array
+import           Data.Hashable
+import           Data.HashMap.Monoidal
+import qualified Data.Map                                           as Map
+import           Data.Text.Prettyprint.Doc
+
+{- Note [Scoping]
+   The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
+   prerequisite. The CEK machine correctly handles name shadowing.
+-}
+
+type TermWithMem uni = Term Name uni ExMemory
+
+{- Note [Arities in VBuiltin]
+The VBuiltin value below contains two copies of the arity (list of
+TypeArg/TermArg pairs) for the relevant builtin.  The second of these
+is consumed as the builtin is instantiated and applied to arguments,
+to check that type and term arguments are interleaved correctly.  The
+first copy of the arity is left unaltered and only used by
+dischargeCekValue if we have to convert the frame back into a term
+(see mkBuiltinApplication).  An alternative would be to look up the
+full arity in mkBuiltinApplication, but that would require a lot of
+things to be monadic (including the PrettyBy instance for CekValue,
+which is a problem.)
+-}
+
+-- 'Values' for the modified CEK machine.
+data CekValue uni =
+    VCon ExMemory (Some (ValueOf uni))
+  | VDelay ExMemory (TermWithMem uni) (CekValEnv uni)
```

Comment:
I was slightly surprised to see that this has an environment, but it makes sense: we store an environment not when we hit a binder, but when we suspend evaluation.